### DISCUSSION TITLE
How to build sdk from source for new architecture

### DISCUSSION THREAD

**Discussion by SzpejnaDawid** [dotnet/source-build#4501] [Q&A]
I would like to ask for any hint how to build SDK for a new architecture (`riscv`). I was trying
```bash
./build.sh --pack -c Release -sb /p:Architecture=riscv64
```
But building failed with error:
```
/dotnet/sdk/artifacts/sb/src/src/Installer/redist-installer/targets/GenerateLayout.targets(404,5): error : 
Download from all targets failed. List of attempted targets: 
https://dotnetcli.blob.core.windows.net/dotnet/Runtime/9.0.0-preview.5.24306.7/dotnet-runtime-9.0.0-preview.5.24306.7-linux-riscv64.tar.gz, 
https://dotnetbuilds.blob.core.windows.net/public/Runtime/9.0.0-preview.5.24306.7/dotnet-runtime-9.0.0-preview.5.24306.7-linux-riscv64.tar.gz
[/dotnet/sdk/artifacts/sb/src/src/Installer/redist-installer/redist-installer.csproj]
```
I was trying to build [dotnet/sdk](https://github.com/dotnet/sdk)

**Comment by omajid** [comment]
If you want to build in source-only mode, please use the `--source-build` (or `--source-only`, an alias) flag. Otherwise the build will try and download all sorts of things from the Microsoft distribution of .NET, many of which will not exist for `riscv64`.

For cross-compilation, I generally use the following additional arguments to `build.sh`:  `/p:TargetArchitecture=ppc64le /p:PortableBuild=true`. This is in an environment with all the required cross-building dependencies installed.

Edit: And use https://github.com/dotnet/dotnet/ for building

**Comment by MichaelSimons** [comment]
For clarity what repo are you building from?

**Reply by SzpejnaDawid** [reply]
I should write it. I was trying to build [dotnet/sdk](https://github.com/dotnet/sdk)

**Reply by omajid** [reply]
Ah, I was using https://github.com/dotnet/dotnet/ in my examples.

**Reply by SzpejnaDawid** [reply]
I was wondering about `dotnet/dotnet` but i didn't see any documentation how to use it and i had a problem with `--with-packages` flag because i am not sure what it means

**Reply by MichaelSimons** [reply]
If you want to source-build the product, you will need to use https://github.com/dotnet/dotnet/.  The -sb option when building an individual repo like the sdk will only build that repo in source-build mode, not all of it's dependencies.  Building individual repos in SB mode is primarily meant to validate the repo's ability to build in the full product source-build environment.  Hopefully that makes sense :)

**Reply by MichaelSimons** [reply]
If you haven't already, please read our documentation on [Building for New Architecture](https://github.com/dotnet/source-build/blob/main/Documentation/bootstrapping-guidelines.md#building-for-new-architecture-using-a-rid-unknown-to-net).  Some of the specifics are out of date given the recent changes to source-build.

**Reply by SzpejnaDawid** [reply]
Ah, okeey. So, may i ask what to do with this information when i am trying to build `dotnet/dotnet` from sources:
```
./build.sh -c Release --source-build /p:TargetArchitecture=riscv64 /p:PortableBuild=true
Cannot find PackagesVersions.props.  Debugging info:
  Attempted custom PVP path: /PackageVersions.props
  Attempted previously-source-built path: /dotnet/prereqs/packages/previously-source-built/PackageVersions.props
  Attempted archive path: /dotnet/prereqs/packages/archive/
```
Should i use `--rid` flag and what with `--with-packages` flag. Do i think correctly that above information is related with that flag?

**Reply by ellahathaway** [reply]
Have you run `./prep-source-build.sh`? This will download a .NET SDK and a number of .NET packages needed to build .NET from source. It also removes binaries disallowed by source-build from `dotnet/dotnet`.

If you don't want to download an SDK or packages, you can pass `--no-sdk` and/or `--no-artifacts` to the prep script. If you wish to provide your own packages to build dotnet, run `./prep-source-build.sh --no-artifacts` and pass `--with-packages <path-to-packages>` to your build command. Similarly, if you wish to provide your own sdk, run `./prep-source-build.sh --no-sdk` and pass `--with-sdk <path-to-sdk>` to your build command.

**Reply by SzpejnaDawid** [reply]
@ellahathaway  Yes, I was trying to use `./prep-source-build.sh` but i see that i did it in a wrong way (I used wrong flags). Now i run
```
./prep-source-build.sh
# below flags may be not minimalist set
./build.sh -c Release  --source-build --rid linux-riscv64 /p:PortableBuild=true /p:TargetArchitecture=riscv64
```
and after some small fixes in `aspnetcore` the compilation ended with an error during `sdk` compilation:

> /dotnet/src/sdk/src/Installer/redist-installer/projects/DownloadPackage.csproj : error NU1100: Unable to resolve 'runtime.linux-riscv64.Microsoft.DotNet.ILCompiler (= 9.0.0-preview.7.24357.2)' for 'net9.0'. PackageSourceMapping is enabled, the following source(s) were not considered: prebuilt, previously-source-built, reference-packages, source-built-aspire, source-built-aspnetcore, source-built-command-line-api, source-built-deployment-tools, source-built-diagnostics, source-built-emsdk, source-built-fsharp, source-built-msbuild, source-built-nuget-client, source-built-roslyn, source-built-roslyn-analyzers, source-built-runtime, source-built-source-build-externals, source-built-sourcelink, source-built-symreader, source-built-templating, source-built-test-templates, source-built-transport-arcade, source-built-transport-aspnetcore, source-built-transport-cecil, source-built-transport-emsdk, source-built-transport-msbuild, source-built-transport-razor, source-built-transport-roslyn, source-built-transport-runtime, source-built-transport-xdt, source-built-vstest.

I noticed that `runtime` doesn't generate `runtime.linux-riscv64.Microsoft.DotNet.ILCompiler ` due to `dotnet/src/runtime/end/Subsets.props`. If I modify `Subsets.props` to build that library, will it work? TBH, i don't see any relation between generating this library locally if the problem is with downloading packages.

**Reply by omajid** [reply]
Do you know if ILCompiler is implemented for riscv64? If it is, then the runtime build system should be taught to build it by default (that may or may not involve Subsets.props). Otherwise we should teach sdk repo to stop requiring it for riscv64.

I think the error message is saying that the package can't be found locally. In `--source-build` mode, nothing should be downloaded by `build.sh` and there shouldn't be any errors related to downloading anything.

**Reply by SzpejnaDawid** [reply]
`runtime` for `riscv64` only supports crossgen2 and there is no support for AOT. 
`runtime.linux-riscv64.Microsoft.DotNet.ILCompiler` is compiled if target architecture supports AOT (`NativeAotSupported` is `true` in eng/Subsets.props). So I have to find a way how to teach sdk repo to stop requiring that lib for `riscv64`...

**Reply by omajid** [reply]
Possibly here: https://github.com/dotnet/sdk/blob/5d2b7f6b3438cb5c42d7b4cdfaccd5fd5603cd7b/src/Installer/redist-installer/redist-installer.csproj#L12

**Comment by sec** [comment]
Hey,
How to trigger cross compilation from VMR? I tried to look at the docs etc. but couldn't find anything related to that. I've tried something like `ROOTFS_DIR=/home/linux/x64_13_rootfs ./build.sh -sb --clean-while-building --rid freebsd.13-x64 /p:TargetArchitecture=x64 /p:PortableBuild=true /p:TargetOS=freebsd` but this produced Linux binaries (as this was the host OS for the build) ?

edit: tried with some extra parameters, like `runtime` is adding those when doing cross build `ROOTFS_DIR=/home/linux/x64_13_rootfs ./build.sh -sb --clean-while-building --rid freebsd.13-x64 /p:TargetArchitecture=x64 /p:PortableBuild=True /p:TargetOS=freebsd /p:CrossBuild=True /p:TargetOsName=freebsd /p:CrossgenOutput=false /p:OSName=freebsd /p:DISABLE_CROSSGEN=True` but this failed with some linker errors... while doing manual runtime build with the same cross parameters is passing fine.

**Reply by omajid** [reply]
Poking around some scripts I use, setting `CROSSCOMPILE=1` and `PKG_CONFIG_PATH=/home/linux/x64_13_rootfs/pkgconfig` might help?

**Reply by sec** [reply]
Nope, looks like VMR build is still doing runtime for linux...

**Comment by omajid** [comment]
Slightly off-topic, can you help me set up a debian:sid (or any other) container that supports cross-compilation from x86_64/amd64 to riscv64?

The obvious thing (at least to me) doesn't work:

```dockerfile
ROM debian:unstable

ARG TARGET_DEBIAN_ARCH
ARG TARGET_GNU_ARCH

ENV TARGET_DEBIAN_ARCH=${TARGET_DEBIAN_ARCH:-riscv64}
ENV TARGET_GNU_ARCH=${TARGET_GNU_ARCH:-riscv64}

ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt-get install -y \
        clang-16 \
        cmake \
        curl \
        g++ \
        gcc \
        git \
        libexpat1-dev \
        libgmp-dev \
        libncurses5-dev \
        libpython3-dev \
        libssl-dev \
        locales \
        make \
        nodejs \
        python3 \
        sudo \
        texinfo \
        wget


RUN dpkg --add-architecture "$TARGET_DEBIAN_ARCH" && \
    apt-get update && \
    apt-get install --no-install-recommends --yes \
        binutils-multiarch \
        crossbuild-essential-"$TARGET_DEBIAN_ARCH" 
```

**Reply by SzpejnaDawid** [reply]
I also have `qemu` and set `/sbin/update-binfmts --enable qemu-riscv64`. And `dotnet/dotnet` requires to have `file` program and `nodejs` and `npm`. However, I don't know if it is all what you need

**Reply by omajid** [reply]
In my case, I can't even install `crossbuild-essential-riscv64` using the dockerfile above. Do you know of any resources that describe how to set up a debian environment for cross-compiling to riscv64?

**Reply by sirntar** [reply]
@omajid At Samsung, we use Ubuntu for cross-compilation. I know it's not Debian, but configuration should be similar. 
  There are two ways: using the official MS docker or creating your own. 

### MS image
As for the official MS image, you can find it here: `mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-22.04-cross-riscv64` It contains all the tools needed for cross-compilation.

### Custom one
As for custom, we use a Dockerfile based on `ubuntu:noble`. I have permission from @bartlomiejko to share part of our setup.

<details>
<summary>ubuntu:noble Dockerfile for RISC-V cross compilation</summary>

```Dockerfile
FROM ubuntu:noble

RUN apt update; apt -y install apt-utils
RUN DEBIAN_FRONTEND=noninteractive apt -y install tzdata

RUN DEBIAN_FRONTEND=noninteractive apt -y install ca-certificates wget curl sudo
RUN wget -qO- https://apt.llvm.org/llvm-snapshot.gpg.key | tee /etc/apt/trusted.gpg.d/apt.llvm.org.asc
RUN echo "deb http://apt.llvm.org/noble/ llvm-toolchain-noble-18 main" >> "/etc/apt/sources.list"
RUN echo "deb-src http://apt.llvm.org/noble/ llvm-toolchain-noble-18 main" >> "/etc/apt/sources.list"
RUN DEBIAN_FRONTEND=noninteractive apt-get update
RUN DEBIAN_FRONTEND=noninteractive apt-get -y install \
 clang-18 \
 llvm-18 \
 lldb-18 \
 lld-18

RUN DEBIAN_FRONTEND=noninteractive apt -y install \
 gcc\
 g++ \
 git \
 libxml2-utils \
 rpm2cpio \
 cpio \
 gcc-arm-linux-gnueabi \
 g++-arm-linux-gnueabi \
 gcc-arm-linux-gnueabihf \
 g++-arm-linux-gnueabihf \
 gcc-aarch64-linux-gnu \
 g++-aarch64-linux-gnu \
 gcc-riscv64-linux-gnu \
 g++-riscv64-linux-gnu \
 vim \
 cmake \
 liblttng-ust-dev \
 locales \
 libicu-dev \
 zlib1g \
 zlib1g-dev \
 libkrb5-dev \
 libssl-dev \
 qemu-user-static \
 debootstrap \
 gdb-multiarch \
 inetutils-ping \
 strace
RUN locale-gen en_US.UTF-8

RUN update-alternatives --install /usr/bin/clang clang /usr/bin/clang-18 10
RUN update-alternatives --install /usr/bin/lldb lldb /usr/bin/lldb-18 10
RUN update-alternatives --install /usr/bin/lld lld /usr/bin/lld-18 10

RUN DEBIAN_FRONTEND=noninteractive apt -y install time

RUN tee /etc/sudoers.d/dotnet-cli <<-_EOF_
Cmnd_Alias DOTNET_ALIAS = /runtime/artifacts/bin/testhost/net9.0-linux-Checked-riscv64/dotnet, \
                         /runtime/artifacts/bin/testhost/net9.0-linux-Release-riscv64/dotnet, \
                         /runtime/artifacts/bin/testhost/net9.0-linux-Debug-riscv64/dotnet

Defaults!DOTNET_ALIAS   env_keep += "QEMU_LD_PREFIX", \
                        env_keep += "qemu_ld_prefix", \
                        env_keep += "ROOTFS_DIR", \
                        env_keep += "rootfs_dir", \
                        env_keep += "DOTNET_EnableWriteXorExecute"

_EOF_

ENV TMPDIR=/tmp

VOLUME [ "/runtime" ]
WORKDIR /runtime
```
```bash
# build docker image
docker build . -t riscv-net
 
# start docker
docker run -ti \
  -v /home/"$USER"/runtime:/runtime \
  --cap-add SYS_PTRACE \
  --name riscross \
  --privileged \
  --sysctl net.ipv6.conf.all.disable_ipv6=0 \
  --sysctl net.ipv6.conf.default.disable_ipv6=0 \
  --hostname $(hostname) \
  riscv-net bash
```
</details>

`/runtime` is folder with cloned `dotnet/runtime` repo.  
After starting docker there is one more thing to be done: configure `crossrootfs` - it will be filesystem using by `qemu` while executing risc-v applications:

<details>
<summary>build crossrootfs</summary>

```bash
# inside docker:
export ARCH=riscv64
export ROOTFS_DIR="/crossrootfs/$ARCH"
export QEMU_LD_PREFIX="$ROOTFS_DIR/"

mkdir -p $ROOTFS_DIR
# build crossrootfs
/runtime/eng/common/cross/build-rootfs.sh $ARCH noble
# enable qemu in kerner - probably require reboot of container
/usr/sbin/update-binfmts --enable qemu-$ARCH

# patch needed for some runtime tests:
cp -r $QEMU_LD_PREFIX/lib/riscv64-linux-gnu /lib
ln -s /lib/riscv64-linux-gnu/ld-linux-riscv64-lp64d.so.1 /lib/ld-linux-riscv64-lp64d.so.1
ln -s /tmp $QEMU_LD_PREFIX/tmp
ln -s /proc $QEMU_LD_PREFIX/proc
```
</details>

**Reply by sirntar** [reply]
And yes, I recommend using the newer `clang`, rather than `clang-16`. The official dotnet compilations use `clang-18` if I remember correctly.

**Reply by omajid** [reply]
Thank you for the pointers!

**Comment by SzpejnaDawid** [comment]
I build `dotnet/dotnet` for `riscv64`. Now i am wondering what changes should be a part of upstream and i have a problem with one thing. When I am building `aspnetcore` there is a problem with `Microsoft.NETCore.App.Crossgen2.linux-x64`'s version. 
```
...
/dotnet/src/aspnetcore/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj : error NU1102: Unable to find package Microsoft.NETCore.App.Crossgen2.linux-x64 with version (>= 9.0.0-preview.7.24363.5) [/dotnet/.dotnet/sdk/9.0.100-preview.7.24357.2/NuGet.targets]
/dotnet/src/aspnetcore/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj : error NU1102:   - Found 1 version(s) in previously-source-built [ Nearest version: 9.0.0-preview.7.24352.15 ] [/dotnet/.dotnet/sdk/9.0.100-preview.7.24357.2/NuGet.targets]
...
```
I don't know what to do with it? I can turn off `crossgen2` for `riscv64` but my colleagues told me that the runtime for riscv should support crossgen2. Moreover, it doesn't look like a problem with `riscv64`. Could anybody help me?

**Reply by ellahathaway** [reply]
This seems very similar to https://github.com/dotnet/sdk/pull/40843#issuecomment-2125740662. Essentially, runtime built a version of crossgen with a different package name than aspnetcore expects. This might be a question for someone on the aspnetcore or runtime team.

cc @wtgodbe @ViktorHofer

**Reply by ViktorHofer** [reply]
@jkoritzinsky / @agocke / @MichalStrehovsky for the crossgen2 question

**Reply by omajid** [reply]
Strangely enough, when I try crossbuilding for `riscv64-linux`, aspnetcore finds just-built crossgen from runtime, but fails when tries to use it:

```
      Running crossgen with /dotnet/src/aspnetcore/artifacts/sb/package-cache/microsoft.netcore.app.crossgen2.linux-x64/9.0.0-preview.7.24365.1/tools/crossgen2
      Error: Unable to load shared library 'clrjit_unix_riscv64_x64' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: 
      /dotnet/src/aspnetcore/artifacts/sb/package-cache/microsoft.netcore.app.crossgen2.linux-x64/9.0.0-preview.7.24365.1/tools/clrjit_unix_riscv64_x64.so: cannot open shared object file: No such file or directory
      /dotnet/src/aspnetcore/artifacts/sb/package-cache/microsoft.netcore.app.crossgen2.linux-x64/9.0.0-preview.7.24365.1/tools/libclrjit_unix_riscv64_x64.so: cannot open shared object file: No such file or directory
      /dotnet/src/aspnetcore/artifacts/sb/package-cache/microsoft.netcore.app.crossgen2.linux-x64/9.0.0-preview.7.24365.1/tools/clrjit_unix_riscv64_x64: cannot open shared object file: No such file or directory
      /dotnet/src/aspnetcore/artifacts/sb/package-cache/microsoft.netcore.app.crossgen2.linux-x64/9.0.0-preview.7.24365.1/tools/libclrjit_unix_riscv64_x64: cannot open shared object file: No such file or directory
```

**Reply by omajid** [reply]
After digging a bit, `libclrjit_unix_riscv64_x64` is the clrjit library used by crossgen that runs on x64 and produces code for riscv64. ASP.NET Core is epxecting this to exist in the microsoft.netcore.app.crossgen2.linux-x64 nuget package. Is this a good assumption? Should `libclrjit_unix_riscv64_x64` be produced and shipped in that nuget package?

**Comment by SzpejnaDawid** [comment]
In the meantime, I would like to ask about one bug that has accompanied me during my `dotnet/dotnet` journey. Sometimes i got this
```
/dotnet/sdk/9.0.100-preview.7.24361.1/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.FrameworkReferenceResolution.targets(158,5): 
error NETSDK1084: There is no application host available for the specified RuntimeIdentifier 'linux-riscv64'
```
and I wasn't able to find the reason behind it. Could anyone explain me this? Thx :D

**Reply by omajid** [reply]
"application host" is what .NET calls the platform-architecture-specific executable that's used to start .NET applications. It sounds like something is trying to publish an application for `linux-riscv64` but the SDK doesn't know about it? Can you share more logs? Maybe a binlog?

**Comment by sec** [comment]
Hi.
Did anyone tried to source build preview.7 on not officialy supported arch? Like building under FreeBSD with prebuilt SDK fails right at the start with
```
  Determining projects to restore...
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103: Unable to find a stable package Microsoft.Build.Tasks.Core with version (>= 17.8.3)
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103:   - Found 6 version(s) in /usr/home/sec/freebsd-ports/lang/dotnet/work/packages/ [ Nearest version: 17.10.0-preview-24076-03 ]
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103:   - Found 5 version(s) in /usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/prereqs/packages/reference/ [ Nearest version: 17.4.0 ]
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103:   - Found 0 version(s) in /usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/prereqs/packages/prebuilt/
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103: Unable to find a stable package Microsoft.Build.Utilities.Core with version (>= 17.8.3)
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103:   - Found 7 version(s) in /usr/home/sec/freebsd-ports/lang/dotnet/work/packages/ [ Nearest version: 17.10.0-preview-24076-03 ]
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103:   - Found 6 version(s) in /usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/prereqs/packages/reference/ [ Nearest version: 17.4.0 ]
/usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj : error NU1103:   - Found 0 version(s) in /usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/prereqs/packages/prebuilt/
  Failed to restore /usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver/Microsoft.DotNet.UnifiedBuild.MSBuildSdkResolver.csproj (in 413 ms).
```
passing `--online` don't make any difference ?

edit: change in `/eng/Versions.props` for correct nuget from bootstrapped SDK that I used helped (I'm using preview.2), but after that there are tons of other errors like
```
  /usr/home/sec/freebsd-ports/lang/dotnet/work/dotnet-9.0.0-preview.7.24405.7/src/source-build-reference-packages/src/targetPacks/ILsrc/netstandard.library/1.6.1/netstandard.library.csproj : error NU1603: netstandard.library depends on runtime.freebsd-x64.Microsoft.NETCore.ILAsm (>= 6.0.0-preview.6.21352.12) but runtime.freebsd-x64.Microsoft.NETCore.ILAsm 6.0.0-preview.6.21352.12 was not found. An approximate best match of runtime.freebsd-x64.Microsoft.NETCore.ILAsm 9.0.0-preview.2.24080.1 was resolved. [/usr/home/sec/freebsd-ports/lang/dotnet/work/bootstrap_sdk/sdk/9.0.100-preview.2.24103.2/NuGet.targets] 
```

**Comment by DancingCorpse** [comment]
Hi, how to build .net core 8 sdk in source-only mode? 
I've cloned the repository "dotnet/dotnet" and checkout the branch "release/8.0.1xx", and running the following command gives the error "_unrecognized argument '-sb'_".
`./build.sh -sb --clean-while-building`

**Reply by mthalman** [reply]
The `-sb` argument was introduced for 9.0. It doesn't apply to 8.0. See the 8.0 branch's README for build instructions: https://github.com/dotnet/dotnet/tree/release/8.0.1xx?tab=readme-ov-file#building. If you need further help, post a separate discussion as this discussion is specific to building with a target architecture.

**Comment by omajid** [comment]
After setting up the right cross-compilation environment, I was able to apply this patch:

```diff
--- a/src/aspnetcore/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj
+++ b/src/aspnetcore/src/Framework/App.Runtime/src/Microsoft.AspNetCore.App.Runtime.csproj
@@ -42,7 +42,7 @@
     <AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>$(AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder);.r2rmap</AllowedOutputExtensionsInSymbolsPackageBuildOutputFolder>

     <!-- Optimize the framework using the crossgen2 tool. Crossgen2 is not currently supported on s390x or ppc64le or armv6 or loongarch64. -->
-    <CrossgenOutput Condition=" '$(TargetArchitecture)' == 's390x' OR '$(TargetArchitecture)' == 'armv6' OR '$(TargetArchitecture)' == 'ppc64le' OR '$(TargetArchitecture)' == 'loongarch64' ">false</CrossgenOutput>
+    <CrossgenOutput Condition=" '$(TargetArchitecture)' == 's390x' OR '$(TargetArchitecture)' == 'armv6' OR '$(TargetArchitecture)' == 'ppc64le' OR '$(TargetArchitecture)' == 'loongarch64' OR '$(TargetArchitecture)' == 'riscv64' ">false</CrossgenOutput>
     <!-- VMR doesn't produce valid crossgen for the host/build machine https://github.com/dotnet/source-build/issues/3793 -->
     <CrossgenOutput Condition=" '$(DotNetBuild)' == 'true' AND '$(Crossbuild)' == 'true' ">false</CrossgenOutput>
     <CrossgenOutput Condition=" '$(CrossgenOutput)' == '' AND '$(Configuration)' != 'Debug' ">true</CrossgenOutput>
--- a/src/sdk/src/Installer/redist-installer/redist-installer.csproj
+++ b/src/sdk/src/Installer/redist-installer/redist-installer.csproj
@@ -9,7 +9,7 @@
     <ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>none</ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>
     <BundleRuntimePacks Condition="'$(BundleRuntimePacks)' == '' and '$(DotNetBuildSourceOnly)' == 'true'">true</BundleRuntimePacks>
     <!-- DotNetBuildOrchestrator is (currently) needed in order to obtain NuGet packages from the runtime build. -->
-    <BundleNativeAotCompiler Condition="'$(BundleNativeAotCompiler)' == '' and '$(DotNetBuildSourceOnly)' == 'true' and '$(SourceBuildUseMonoRuntime)' != 'true' and '$(DotNetBuildOrchestrator)' == 'true'">true</BundleNativeAotCompiler>
+    <BundleNativeAotCompiler Condition="'$(BundleNativeAotCompiler)' == '' and '$(DotNetBuildSourceOnly)' == 'true' and '$(SourceBuildUseMonoRuntime)' != 'true' and '$(DotNetBuildOrchestrator)' == 'true' and '$(Architecture)' != 'riscv64'">true</BundleNativeAotCompiler>
   </PropertyGroup>

   <ItemGroup>
```

And then build it using the following commands:

```
$ export PKG_CONFIG_PATH=/usr/lib/riscv64-linux-gnu/pkgconfig
$ export ROOTFS_DIR=/
$ export CROSSCOMPILE=1
$ export DISABLE_CROSSGEN=true
$ ./prep-source-build.sh
$ ./build.sh --source-only --online /p:TargetArchitecture=riscv64 /p:PortableBuild=true
$ find dotnet -iname 'dotnet-sdk-*tar.gz'
dotnet/src/sdk/artifacts/packages/Release/Shipping/dotnet-sdk-10.0.100-alpha.2.24430.1-linux-riscv64.tar.gz
dotnet/artifacts/assets/Release/Sdk/10.0.100-alpha.2.24430.1/dotnet-sdk-10.0.100-alpha.2.24430.1-linux-riscv64.tar.gz
dotnet/artifacts/assets/Release/dotnet-sdk-10.0.100-alpha.2.24430.1-linux-riscv64.tar.gz
```
